analyzers/src/SonarAnalyzer.CSharp/Rules/AssertionArgsShouldBePassedInCorrectOrder.cs

```diff
@@ -18,63 +18,50 @@
+{
+    internal const string DiagnosticId = "S3415";
+    private const string MessageFormat = "Make sure these 2 arguments are in the correct order: expected value, actual value.";
 
-                var methodCallExpression = (MemberAccessExpressionSyntax)methodCall.Expression;
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-                var methodKnownTypes = MethodsWithType.GetValueOrDefault(methodCallExpression.Name.Identifier.ValueText);
-                if (methodKnownTypes == null)
-                {
-                    return;
-                }
+    private static readonly IDictionary<string, ImmutableArray<KnownType>> MethodsWithType = new Dictionary<string, ImmutableArray<KnownType>>
+    {
+        ["AreEqual"]    = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["AreNotEqual"] = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["AreSame"]     = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["AreNotSame"]  = ImmutableArray.Create(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert, KnownType.NUnit_Framework_Assert),
+        ["Equal"]       = ImmutableArray.Create(KnownType.Xunit_Assert),
+        ["Same"]        = ImmutableArray.Create(KnownType.Xunit_Assert),
+        ["NotSame"]     = ImmutableArray.Create(KnownType.Xunit_Assert)
+    };
 
-                var symbolInfo = c.SemanticModel.GetSymbolInfo(methodCallExpression.Expression).Symbol;
-                var isAnyTrackedAssertType = (symbolInfo as INamedTypeSymbol).IsAny(methodKnownTypes);
-                if (!isAnyTrackedAssertType)
-                {
-                    return;
-                }
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+        {
+            if (c.Node is InvocationExpressionSyntax { ArgumentList: { Arguments: { Count: >= 2 } arguments } argumentList, Expression: MemberAccessExpressionSyntax methodCallExpression }
+                && methodCallExpression.IsKind(SyntaxKind.SimpleMemberAccessExpression)
+                && CSharpFacade.Instance.MethodParameterLookup(argumentList, c.SemanticModel) is var parameterLookup
+                && CheckArguments(parameterLookup)
+                && MethodsWithType.GetValueOrDefault(methodCallExpression.Name.Identifier.ValueText) is { } methodKnownTypes
+                && methodKnownTypes.IsDefault is false
+                && (c.SemanticModel.GetSymbolInfo(methodCallExpression.Expression).Symbol as INamedTypeSymbol).IsAny(methodKnownTypes))
```

Comment:
This more robust as `methodCallExpression.Expression` is tricky. E.g. for instance methods it works for `A.B.C()` (returns B for the C() invocation) but probably fails for `A?.B?.C()`.
```suggestion
                && c.SemanticModel.GetSymbolInfo(invocation) is { Symbol: IMethodSymbol { IsStatic: true, ContainingSymbol: { IsStatic: true } container } 
                && container.IsAny(methodKnownTypes)
```

Reply:
I don't understand the intend of this change. What cases are you trying to catch here? I tried it anyway but most tests failed:
```
&& c.SemanticModel.GetSymbolInfo(invocation) is { Symbol: IMethodSymbol { IsStatic: true, ContainingSymbol: INamedTypeSymbol{ IsStatic: true } container } }
&& container.IsAny(methodKnownTypes))
```